File: packages/core/src/algorithms/shortest-path.ts
import { ErrorCode, LenkeError } from '@lenke/errors';

import type { Edge } from '../core/Edge.js';
import type { Graph } from '../core/Graph.js';
import type { Vertex } from '../core/Vertex.js';
import { incidentEdges } from './adjacency.js';
import { type AlgorithmGen, defineAlgorithm, materializeVertices, YIELD_EVERY } from './async.js';
import type { AlgorithmConfig, AlgorithmRow } from './types.js';

/** A shortest-path result row: `{ node, distance }`. */
export type ShortestPathRow = AlgorithmRow<'distance', number>;

/** A binary min-heap entry: `[distance, vertexIndex]`, ordered by distance then index. */
type HeapItem = [number, number];

const less = (a: HeapItem, b: HeapItem): boolean => a[0] < b[0] || (a[0] === b[0] && a[1] < b[1]);

const heapPush = (heap: HeapItem[], item: HeapItem): void => {
  heap.push(item);
  let i = heap.length - 1;

  while (i > 0) {
    const parent = (i - 1) >> 1;

    if (!less(heap[i], heap[parent])) {
      break;
    }

    [heap[i], heap[parent]] = [heap[parent], heap[i]];
    i = parent;
  }
};

const heapPop = (heap: HeapItem[]): HeapItem => {
  const [top] = heap;
  const last = heap.pop()!;

  if (heap.length > 0) {
    heap[0] = last;
    let i = 0;

    for (;;) {
      const l = 2 * i + 1;
      const r = l + 1;
      let smallest = i;

      if (l < heap.length && less(heap[l], heap[smallest])) {
        smallest = l;
      }

      if (r < heap.length && less(heap[r], heap[smallest])) {
        smallest = r;
      }

      if (smallest === i) {
        break;
      }

      [heap[i], heap[smallest]] = [heap[smallest], heap[i]];
      i = smallest;
    }
  }

  return top;
};

/**
 * Yield each out-edge from a vertex's `edgesFromByLabel` entry, optionally
 * restricted to one edge type. The caller reads `edge.to` for the neighbour.
 */
const outEdges = incidentEdges;

/** Which incident edges to follow, from `config.direction` (default `out`, mirroring
 *  `degree`); `both` treats the graph as undirected. */
type Dir = 'out' | 'in' | 'both';
const dirOf = (config: AlgorithmConfig): Dir =>
  config.direction === 'in' || config.direction === 'both' ? config.direction : 'out';

/**
 * Yield `[edge, neighbourId]` for each incident edge of `u` in the configured
 * direction — the far endpoint is `edge.to` for an out-edge, `edge.from` for an
 * in-edge. `both` yields out-edges then in-edges, matching native `visit_adj`'s
 * order so the Dijkstra `(distance, index)` tie-break stays byte-identical.
 */
const neighbours = function* (
  graph: Graph,
  u: Vertex,
  dir: Dir,
  edgeLabel: string | undefined,
): Iterable<[Edge, string]> {
  if (dir === 'out' || dir === 'both') {
    for (const e of outEdges(graph.edgesFromByLabel.get(u.id), edgeLabel)) {
      yield [e, e.to.id];
    }
  }

  if (dir === 'in' || dir === 'both') {
    for (const e of outEdges(graph.edgesToByLabel.get(u.id), edgeLabel)) {
      yield [e, e.from.id];
    }
  }
};

/**
 * Goal-directed A* — explore by `f = g + h`, where `h` is the admissible estimate
 * to the target read from each vertex's `heuristicProperty` (absent → 0, degrading
 * to Dijkstra). Returns the source→target distance (optimal when the target is
 * settled, so identical to Dijkstra), or `undefined` if unreachable. Same
 * `(priority, index)` tie-break as the Dijkstra heap, so native and TS agree.
 */
/** Shared traversal context — the graph plus the derived index/weight helpers. */
type PathContext = {
  graph: Graph;
  order: Vertex[];
  index: Map<string, number>;
  edgeLabel: string | undefined;
  dir: Dir;
  weightOf: (edge: Edge) => number;
};

const astar = (
  ctx: PathContext,
  srcIdx: number,
  tgtIdx: number,
  heuristicProperty: string | undefined,
): number | undefined => {
  const { graph, order, index, edgeLabel, dir, weightOf } = ctx;
  const heuristicOf = (i: number): number => {
    if (heuristicProperty === undefined) {
      return 0;
    }

    const h = order[i].getProperty(heuristicProperty);

    return typeof h === 'number' ? h : 0;
  };

  const g = new Float64Array(order.length).fill(Infinity);
  const closed = new Uint8Array(order.length);
  g[srcIdx] = 0;
  const heap: HeapItem[] = [[heuristicOf(srcIdx), srcIdx]];

  while (heap.length > 0) {
    const [, u] = heapPop(heap);

    if (closed[u]) {
      continue;
    }

    closed[u] = 1;

    if (u === tgtIdx) {
      return g[u];
    }

    for (const [edge, vid] of neighbours(graph, order[u], dir, edgeLabel)) {
      const v = index.get(vid)!;

      if (closed[v]) {
        continue;
      }

      const ng = g[u] + weightOf(edge);

      if (ng < g[v]) {
        g[v] = ng;
        heapPush(heap, [ng + heuristicOf(v), v]);
      }
    }
  }

  return undefined;
};

export const computeGen = function* (
  config: AlgorithmConfig,
  graph: Graph,
): AlgorithmGen<ShortestPathRow> {
  const { source, target, edgeLabel, weightProperty, heuristicProperty, algorithm, writeProperty } =
    config;

  const order = yield* materializeVertices(graph);
  const index = new Map<string, number>();

  order.forEach((v, i) => index.set(v.id, i));

  const srcIdx = source === undefined ? undefined : index.get(source);
  const dir = dirOf(config);

  // Unknown/absent source → no reachable set.
  if (srcIdx === undefined) {
    return [];
  }

  // Dijkstra (and A*) require NON-NEGATIVE weights. With a negative edge the
  // relaxation can keep finding a cheaper path forever, so a negative cycle never
  // terminates — and a negative self-loop is enough: one node and one edge hung
  // the engine indefinitely. Rejected for ANY negative weight, not just a cycle,
  // because Dijkstra can also settle a vertex too early and return a silently
  // wrong distance; a negative-but-acyclic graph terminating is luck, not a
  // correct answer. Mirrors the native guard in `algo/mod.rs`.
  if (weightProperty !== undefined) {
    for (const edge of graph.edges) {
      const w = edge.getProperty(weightProperty);

      // NaN is rejected alongside negatives: it makes every relaxation
      // comparison false and strands the search just as effectively.
      if (typeof w === 'number' && (Number.isNaN(w) || w < 0)) {
        throw new LenkeError(
          `shortestPath \`weightProperty\` (${weightProperty}) must hold non-negative numbers — Dijkstra does not admit negative weights`,
          { code: ErrorCode.InvalidValue },
        );
      }
    }
  }

  const weightOf = (edge: Edge): number => {
    if (weightProperty === undefined) {
      return 1;
    }

    const w = edge.getProperty(weightProperty);

    return typeof w === 'number' ? w : 0;
  };

  // A* is a goal-directed backend returning just the source→target distance
  // (identical to Dijkstra's), exploring fewer vertices via the admissible
  // heuristic. Falls back to no rows for an unknown/unreachable target.
  if (algorithm === 'astar') {
    const tgtIdx = target === undefined ? undefined : index.get(target);

    if (tgtIdx === undefined) {
      return [];
    }

    const ctx: PathContext = { graph, order, index, edgeLabel, dir, weightOf };
    const d = astar(ctx, srcIdx, tgtIdx, heuristicProperty);

    if (d === undefined) {
      return [];
    }

    if (writeProperty !== undefined) {
      order[tgtIdx].setProperty(writeProperty, d);
    }

    return [{ node: order[tgtIdx].id, distance: d }];
  }

  const dist = new Float64Array(order.length).fill(Infinity);
  dist[srcIdx] = 0;

  let sinceYield = 0;

  if (weightProperty === undefined) {
    // Unweighted BFS — hop distance (order-independent unique layers).
    const queue: number[] = [srcIdx];
    let head = 0;

    while (head < queue.length) {
      const u = queue[head++];

      for (const [, vid] of neighbours(graph, order[u], dir, edgeLabel)) {
        const v = index.get(vid)!;

        if (dist[v] === Infinity) {
          dist[v] = dist[u] + 1;
          queue.push(v);
        }
      }

      if (++sinceYield >= YIELD_EVERY) {
        sinceYield = 0;

        yield;
      }
    }
  } else {
    // Weighted Dijkstra — min-heap keyed by (distance, vertex index). The settled
    // distance is the canonical minimum path cost, so it matches native exactly.
    const heap: HeapItem[] = [[0, srcIdx]];

    while (heap.length > 0) {
      const [du, u] = heapPop(heap);

      if (du > dist[u]) {
        continue;
      }

      for (const [edge, vid] of neighbours(graph, order[u], dir, edgeLabel)) {
        const v = index.get(vid)!;
        const nd = du + weightOf(edge);

        if (nd < dist[v]) {
          dist[v] = nd;
          heapPush(heap, [nd, v]);
        }
      }

      if (++sinceYield >= YIELD_EVERY) {
        sinceYield = 0;

        yield;
      }
    }
  }

  // A `target` restricts the result to that one vertex's distance (like A*),
  // instead of a row per reachable vertex. Unknown/unreachable target → no rows.
  if (target !== undefined) {
    const ti = index.get(target);

    if (ti === undefined || !Number.isFinite(dist[ti])) {
      return [];
    }

    if (writeProperty !== undefined) {
      order[ti].setProperty(writeProperty, dist[ti]);
    }

    return [{ node: order[ti].id, distance: dist[ti] }];
  }

  const rows: ShortestPathRow[] = [];

  order.forEach((v, i) => {
    if (!Number.isFinite(dist[i])) {
      return;
    }

    if (writeProperty !== undefined) {
      v.setProperty(writeProperty, dist[i]);
    }

    rows.push({ node: v.id, distance: dist[i] });
  });

  return rows;
};

/**
 * Single-source shortest path from a `source` external id, following edges in the
 * configured `direction` (`out` default / `in` / `both`; `both` = undirected),
 * optionally of one `edgeLabel`. Unweighted → BFS hop distance; weighted (a
 * `weightProperty` is set) → Dijkstra. Resolves `Promise<ShortestPathRow[]>` for
 * every reachable vertex (source at 0), in insertion order, without blocking the
 * event loop. Distances are the canonical minimum path cost, so they are
 * byte-identical to the native engine. Data-last dual-form: `shortestPath(config,
 * graph)` or `shortestPath(config)(graph)`.
 */
export const shortestPath = defineAlgorithm(computeGen);

File: crates/lenke-core/src/algo/shortest_path.rs
//! Single-source shortest path from a `source` external id, following edges in the
//! configured `direction` (`out` default / `in` / `both`, mirroring `degree`;
//! `both` = undirected), optionally of one type. Unweighted → BFS integer hop
//! distance; weighted (a
//! `weightProperty` is set) → Dijkstra f64 distance. Returns `{node, distance}` for
//! every reachable vertex (including the source at 0), in vertex-insertion order.
//!
//! Cross-engine identity: a vertex's shortest distance is the canonical minimum
//! over all path costs — BFS layer distances are unique integers, and Dijkstra's
//! settled distance is the minimum path float-sum — so both engines produce the
//! same distances. The priority queue breaks ties by (distance, then vertex index)
//! so the *exploration* order is identical too, keeping even float-pathological
//! graphs byte-identical. Unknown/absent source → no rows.

use std::cmp::Ordering;
use std::collections::{BinaryHeap, VecDeque};

use super::AlgoConfig;
use crate::graph::{Adj, Graph, Value};

/// A Dijkstra frontier entry ordered as a min-heap on `(dist, idx)`: `BinaryHeap`
/// is a max-heap, so `Ord` is reversed — the smallest distance (then smallest
/// vertex index) compares greatest and pops first.
struct State {
    dist: f64,
    idx: u32,
}

impl PartialEq for State {
    fn eq(&self, other: &Self) -> bool {
        self.dist == other.dist && self.idx == other.idx
    }
}
impl Eq for State {}
impl Ord for State {
    fn cmp(&self, other: &Self) -> Ordering {
        other
            .dist
            .partial_cmp(&self.dist)
            .unwrap_or(Ordering::Equal)
            .then_with(|| other.idx.cmp(&self.idx))
    }
}
impl PartialOrd for State {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

/// Which incident edges the traversal follows, from `AlgorithmConfig.direction`
/// (default `out`, mirroring `degree`); `both` treats the graph as undirected.
#[derive(Clone, Copy)]
enum Dir {
    Out,
    In,
    Both,
}

impl Dir {
    fn of(cfg: &AlgoConfig) -> Self {
        match cfg.direction.as_deref() {
            Some("in") => Self::In,
            Some("both") => Self::Both,
            _ => Self::Out,
        }
    }
}

/// Visit each incident edge of `u` in the configured direction. `a.nbr` is the far
/// endpoint in every case (target of an out-edge, source of an in-edge), so callers
/// relax toward `a.nbr` uniformly. `both` visits out-edges then in-edges — both
/// engines share this order so Dijkstra's `(dist, idx)` tie-break stays identical.
fn visit_adj(graph: &Graph, u: u32, dir: Dir, f: &mut impl FnMut(&Adj)) {
    match dir {
        Dir::Out => {
            for a in graph.out_adj(u) {
                f(&a);
            }
        }
        Dir::In => {
            for a in graph.in_adj(u) {
                f(&a);
            }
        }
        Dir::Both => {
            for a in graph.out_adj(u) {
                f(&a);
            }
            for a in graph.in_adj(u) {
                f(&a);
            }
        }
    }
}

pub fn shortest_path(graph: &Graph, cfg: &AlgoConfig) -> Vec<(u32, Value)> {
    // Resolve the source external id → dense id; unknown/absent → no reachable set.
    let Some(src) = cfg.source.as_deref().and_then(|s| graph.vid.get(s)) else {
        return Vec::new();
    };
    // A named-but-unknown edge type → only the source is reachable (no edges).
    let etype = cfg.etype(graph);
    // Whether any edge carries the wanted type as a NON-first label, resolved
    // once. When it does not — which is every graph with single-type edges — the
    // adjacency entry alone decides and `graph` is never dereferenced in the
    // walk. Consulting it per edge instead cost 1.08x on a 1M/8M BFS.
    let need_extra = matches!(etype, Some(Some(t)) if graph.etypes_need_extra_lookup(&[t]));
    let passes = |a: &Adj| match etype {
        Some(Some(t)) => a.etype == t || (need_extra && graph.edge_has_label(a.eidx, t)),
        Some(None) => true,
        None => false,
    };

    let slots = graph.n;
    let dir = Dir::of(cfg);
    // Precompute per-edge weights once (weighted runs read them per relaxation, so a
    // hashed property lookup there would dominate). `None` = unweighted.
    let weights: Option<Vec<f64>> = cfg
        .weight_property
        .as_deref()
        .map(|k| super::edge_weights(graph, k));

    // A* is a goal-directed backend: given a `target`, it returns just the source→
    // target distance (identical to Dijkstra's, so interchangeable), exploring far
    // fewer vertices via the admissible `heuristicProperty`.
    if cfg.algorithm.as_deref() == Some("astar") {
        let Some(tgt) = cfg.target.as_deref().and_then(|t| graph.vid.get(t)) else {
            return Vec::new();
        };
        return match astar(graph, src, tgt, slots, cfg, weights.as_deref(), &passes) {
            Some(d) => vec![(tgt, Value::Num(d))],
            None => Vec::new(),
        };
    }

    // Full SSSP (the default): unweighted BFS layers, or weighted Dijkstra.
    let dist = match weights.as_deref() {
        None => bfs(graph, src, slots, dir, &passes),
        Some(w) => dijkstra(graph, src, slots, dir, w, &passes),
    };

    // A `target` restricts the result to that one vertex's distance (like A*),
    // instead of a row per reachable vertex. Unknown/unreachable target → no rows.
    if let Some(t) = cfg.target.as_deref() {
        let Some(tgt) = graph.vid.get(t).filter(|&v| dist[v as usize].is_finite()) else {
            return Vec::new();
        };
        return vec![(tgt, Value::Num(dist[tgt as usize]))];
    }

    graph
        .vertex_indices()
        .filter(|&v| dist[v as usize].is_finite())
        .map(|v| (v, Value::Num(dist[v as usize])))
        .collect()
}

/// Goal-directed A\*: explore by `f = g + h`, where `h` is the admissible estimate
/// to `tgt` read from each vertex's `heuristicProperty` (absent → 0, degrading to
/// Dijkstra). Returns `Some(distance)` when `tgt` is settled (its `g` is then
/// optimal — identical to Dijkstra), `None` if unreachable. Edge weights come from
/// `weightProperty` (absent → unit weights). Same `(priority, idx)` tie-break as
/// Dijkstra, so native and TS explore identically.
fn astar(
    graph: &Graph,
    src: u32,
    tgt: u32,
    slots: usize,
    cfg: &AlgoConfig,
    weights: Option<&[f64]>,
    passes: &impl Fn(&Adj) -> bool,
) -> Option<f64> {
    let dir = Dir::of(cfg);
    let hkey = cfg.heuristic_property.as_deref();
    let h = |v: u32| -> f64 {
        match hkey {
            None => 0.0,
            Some(k) => match graph.props.value(v as usize, k, &graph.strs) {
                Value::Num(x) => x,
                _ => 0.0,
            },
        }
    };
    let weight = |a: &Adj| -> f64 { weights.map_or(1.0, |w| w[a.eidx as usize]) };

    let mut g = vec![f64::INFINITY; slots];
    let mut closed = vec![false; slots];
    g[src as usize] = 0.0;
    let mut heap = BinaryHeap::new();
    heap.push(State {
        dist: h(src),
        idx: src,
    });
    while let Some(State { idx: u, .. }) = heap.pop() {
        if closed[u as usize] {
            continue;
        }
        closed[u as usize] = true;
        if u == tgt {
            return Some(g[u as usize]);
        }
        visit_adj(graph, u, dir, &mut |a| {
            if !passes(a) || closed[a.nbr as usize] {
                return;
            }
            let ng = g[u as usize] + weight(a);
            if ng < g[a.nbr as usize] {
                g[a.nbr as usize] = ng;
                heap.push(State {
                    dist: ng + h(a.nbr),
                    idx: a.nbr,
                });
            }
        });
    }
    None
}

/// Unweighted BFS hop distance (as f64), `INFINITY` for unreached.
fn bfs(
    graph: &Graph,
    src: u32,
    slots: usize,
    dir: Dir,
    passes: &impl Fn(&Adj) -> bool,
) -> Vec<f64> {
    let mut dist = vec![f64::INFINITY; slots];
    dist[src as usize] = 0.0;
    let mut queue = VecDeque::new();
    queue.push_back(src);
    while let Some(u) = queue.pop_front() {
        let du = dist[u as usize];
        visit_adj(graph, u, dir, &mut |a| {
            if passes(a) && dist[a.nbr as usize].is_infinite() {
                dist[a.nbr as usize] = du + 1.0;
                queue.push_back(a.nbr);
            }
        });
    }
    dist
}

/// Weighted Dijkstra f64 distance, `INFINITY` for unreached. `weights` is the
/// precomputed per-edge weight (indexed by edge id); negative weights are out of
/// contract (Dijkstra).
fn dijkstra(
    graph: &Graph,
    src: u32,
    slots: usize,
    dir: Dir,
    weights: &[f64],
    passes: &impl Fn(&Adj) -> bool,
) -> Vec<f64> {
    let mut dist = vec![f64::INFINITY; slots];
    dist[src as usize] = 0.0;
    let mut heap = BinaryHeap::new();
    heap.push(State {
        dist: 0.0,
        idx: src,
    });
    while let Some(State { dist: du, idx: u }) = heap.pop() {
        // Skip a stale entry (a shorter distance was already settled).
        if du > dist[u as usize] {
            continue;
        }
        visit_adj(graph, u, dir, &mut |a| {
            if !passes(a) {
                return;
            }
            let nd = du + weights[a.eidx as usize];
            if nd < dist[a.nbr as usize] {
                dist[a.nbr as usize] = nd;
                heap.push(State {
                    dist: nd,
                    idx: a.nbr,
                });
            }
        });
    }
    dist
}
